### ISSUE TITLE
Get rid of jsbn and crypto-js?

### ISSUE THREAD

**Issue by jbis9051** [midonet/tssrp6a#47]
With Node support for JavaScript's native big integer type, [BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt), could we eliminate the jsbn dependency? 

Node also has a bunch of crypto functions in the [crypto module](https://nodejs.org/api/crypto.html). Could we eliminate the `crypto-js` dependency as well?

I would favor native capabilities rather than third part solutions when possible, unless of course there is something missing from those native solutions.

**Comment by bgrosse-midokura** [comment]
tssrp6a must be able to run in the browser, BigInt compatibility looks quite good. jsbn and BigInt might not behave in completely the same way, though. But I think the current test suite would catch any deviations.

For crypto, node's crypto isn't available on browsers, but there is https://developer.mozilla.org/en-US/docs/Web/API/Crypto which might as well work.

I would definitely favour using BigInt. For crypto, it looks a bit more complicated. Perhaps there is some ready-made module that offers the relevant crypto functions with fallbacks for node.js and browser?

**Comment by jbis9051** [comment]
@bgrosse-midokura  As for the crypto issue, we could utilize `browserify` to polyfill node's crypto library (or webpack, but due to webpack 5 drop for native node module and the simplicity of this project browserify is probably better).

**Comment by bgrosse-midokura** [comment]
I would be reluctant to change the current simple `tsc` build. I think simply using a polyfill module (that has auto-fallback to native) would be sufficient, no?

**Comment by jbis9051** [comment]
My concern with that approach is that the polyfill would be included in the build regardless of it's needed (I think). Node's build size would include the polyfill for no reason. 

Based on [TypeScript's docs](https://www.typescriptlang.org/docs/handbook/integrating-with-build-tools.html#browserify), the build command complexity would only increase slightly. 

```
browserify src/index.ts -p [ tsify --noImplicitAny ] > bundle.js
```

**Comment by bgrosse-midokura** [comment]
It's either included, or there have to be two build artifacts/targets
between which a user then also has to chose. This sounds like too much
difficulty for some size improvement.

Could this https://github.com/PeculiarVentures/webcrypto maybe just work?

**Comment by jbis9051** [comment]
> It's either included, or there have to be two build artifacts/targets
between which a user then also has to chose. This sounds like too much
difficulty for some size improvement.

I agree. Although maybe two builds isn't a terrible idea?

> Could this https://github.com/PeculiarVentures/webcrypto maybe just work?

I'd be concerned about the lack of support for many algorithms as well as the warning that it shouldn't be used in production.

How about something like https://nodejs.org/api/webcrypto.html#webcrypto_web_crypto_api? It's experimental but implements the same API available in the browser. 

We could do something like 

```js
const subtle = crypto.subtle || require('crypto').webcrypto.subtle;
```

**Comment by bgrosse-midokura** [comment]
Yes, that node.js webcrypto looks much better, and is official, although experimental. Sounds like the best solution.

**Comment by bgrosse-midokura** [comment]
@jbis9051 I'm giving native BigInt a try

**Comment by bgrosse-midokura** [comment]
@jbis9051 native BigInt fails with `RangeError: Maximum BigInt size exceeded`. I've opened #51 , you can take a look if you see a way to fix it. I think it boils down to V8.

@bufistov I think for that I just used `.toString(16)` in that PR.

**Comment by bufistov** [comment]
I see, nice! :)

**Comment by jbis9051** [comment]
jsbn removed by https://github.com/midonet/tssrp6a/pull/51

**Comment by jbis9051** [comment]
Looking more into webcrypto. There are less hash functions available (SHA-1, SHA-256, SHA-384 SHA-512), but that will give us an opportunity to implement #53. I'll try to work on a PR.

**Comment by jbis9051** [comment]
Seems webcrypto was added in Node v15 If we want to support <15.0, we could add compatibility with the normal crypto library. Looking into that.
